npe2/_plugin_manager.py

```diff
@@ -52,105 +54,220 @@ def addi(self, begin: int, end: int, data: T) -> None:
+        set,
+        description="Emitted with a set of PluginManifest instances "
+        "whenever new plugins are registered. 'Registered' means that a "
+        "manifest has been provided or discovered.",
+    )
+    activation_changed = Signal(
+        set,
+        set,
+        description="Emitted with two arguments: a set of plugin "
+        "names that were activated, and a set of names that were "
+        "deactivated. 'Activated' means the plugin has been *imported*, it's "
+        "`on_activate` function was called.",
+    )
+    enablement_changed = Signal(
+        set,
+        set,
+        description="Emitted with two arguments: a set of plugin names "
+        "that were enabled, and a set of names that were "
+        "disabled. 'Disabled' means the plugin remains installed, but it "
+        "cannot be activated, and its contributions will not be indexed.",
+    )
 
 
 class PluginManager:
     __instance: Optional[PluginManager] = None  # a global instance
     _contrib: _ContributionsIndex
+    events: PluginManagerEvents
 
-    def __init__(self, reg: Optional[CommandRegistry] = None) -> None:
+    def __init__(
+        self, *, disable: Iterable[str] = (), reg: Optional[CommandRegistry] = None
+    ) -> None:
+        self._disabled_plugins: Set[PluginName] = set(disable)
         self._command_registry = reg or CommandRegistry()
         self._contexts: Dict[PluginName, PluginContext] = {}
+        self._contrib = _ContributionsIndex()
         self._manifests: Dict[PluginName, PluginManifest] = {}
-        self.discover()  # TODO: should we be immediately discovering?
+        self.events = PluginManagerEvents(self)
+
+    @classmethod
+    def instance(cls) -> PluginManager:
+        """Singleton instance."""
+        if cls.__instance is None:
+            cls.__instance = cls()
+        return cls.__instance
 
     @property
     def commands(self) -> CommandRegistry:
         return self._command_registry
 
-    def discover(self, paths: Sequence[str] = ()) -> None:
+    # Discovery, activation, enablement
+
+    def discover(self, paths: Sequence[str] = (), clear=False) -> None:
         """Discover and index plugin manifests in the environment.
 
         Parameters
         ----------
-        paths : Sequence[str], optional
-            Optional list of strings to insert at front of sys.path when discovering.
+        paths : Sequence[str]
+            Optional list of strings to insert at front of sys.path when
+            discovering.
+        clear : bool
+            Clear and re-index the environment.  If `False` (the default),
+            calling discover again will only register and index newly
+            discovered plugins. (Existing manifests will not be re-indexed)
         """
-        self._contrib = _ContributionsIndex()
-        self._manifests.clear()
+        if clear:
+            self._contrib = _ContributionsIndex()
+            self._manifests.clear()
```

Comment:
it's not evented (I think ?) so  you can also just assign an empty dict ? Would it be faster ? Unless something else keep reference to `_manifest` ?

Reply:
i'm trying to keep all private access within classes, even within npe2, so it should be fine just to set to empty dict